File: frontend/src/services/petitionService.js
import axios from 'axios';
import authService from './authService';

const API_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

const getAuthHeader = () => {
    const token = authService.getToken();
    console.log('Token:', token); // Log token
    if (!token) {
        throw new Error('No authentication token found');
    }
    return {
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }
    };
};

const petitionService = {
    /**
     * Lấy danh sách đơn thỉnh cầu của phụ huynh
     * @param {number} parentId - ID của phụ huynh
     * @param {number} page - Số trang
     * @param {number} size - Số lượng đơn trên mỗi trang
     * @returns {Promise<Object>} - Danh sách đơn và thông tin phân trang
     */
    getParentPetitions: async (parentId, page = 1, size = 10) => {
        if (!parentId || isNaN(parentId)) {
            throw new Error('Invalid parentId: parentId must be a valid number');
        }
        try {
            const response = await axios.get(`${API_URL}/petitions/parent/${parentId}`, {
                ...getAuthHeader(),
                params: { page, size }
            });
            return response.data;
        } catch (error) {
            if (error.response?.status === 401) {
                await authService.refreshToken();
                const response = await axios.get(`${API_URL}/petitions/parent/${parentId}`, {
                    ...getAuthHeader(),
                    params: { page, size }
                });
                return response.data;
            }
            throw error.response?.data?.detail || error.message || 'Failed to fetch parent petitions';
        }
    },

    /**
     * Lấy tất cả đơn thỉnh cầu (dành cho admin)
     * @param {Object} filters - Các bộ lọc
     * @param {string} filters.status - Trạng thái đơn
     * @param {number} filters.parentId - ID phụ huynh
     * @param {string} filters.startDate - Ngày bắt đầu
     * @param {string} filters.endDate - Ngày kết thúc
     * @param {number} page - Số trang
     * @param {number} size - Số lượng đơn trên mỗi trang
     * @returns {Promise<Object>} - Danh sách đơn và thông tin phân trang
     */
    getAllPetitions: async (filters = {}, page = 1, size = 10) => {
        try {
            const response = await axios.get(`${API_URL}/petitions/`, {
                ...getAuthHeader(),
                params: { ...filters, page, size }
            });
            return response.data;
        } catch (error) {
            if (error.response?.status === 401) {
                await authService.refreshToken();
                const response = await axios.get(`${API_URL}/petitions/`, {
                    ...getAuthHeader(),
                    params: { ...filters, page, size }
                });
                return response.data;
            }
            throw error.response?.data?.detail || error.message || 'Failed to fetch all petitions';
        }
    },

    /**
     * Lấy thông tin chi tiết một đơn thỉnh cầu
     * @param {number} petitionId - ID của đơn thỉnh cầu
     * @returns {Promise<Object>} - Thông tin chi tiết đơn
     */
    getPetitionById: async (petitionId) => {
        if (!petitionId || isNaN(petitionId)) {
            throw new Error('Invalid petitionId: petitionId must be a valid number');
        }
        try {
            const response = await axios.get(`${API_URL}/petitions/${petitionId}`, getAuthHeader());
            return response.data;
        } catch (error) {
            if (error.response?.status === 401) {
                await authService.refreshToken();
                const response = await axios.get(`${API_URL}/petitions/${petitionId}`, getAuthHeader());
                return response.data;
            }
            throw error.response?.data?.detail || error.message || 'Failed to fetch petition details';
        }
    },

    /**
     * Tạo đơn thỉnh cầu mới
     * @param {Object} petition - Thông tin đơn thỉnh cầu
     * @param {string} petition.Title - Tiêu đề đơn
     * @param {string} petition.Content - Nội dung đơn
     * @param {number} petition.ParentID - ID phụ huynh
     * @returns {Promise<Object>} - Đơn thỉnh cầu đã được tạo
     */
    createPetition: async (petition) => {
        if (!petition.ParentID || isNaN(petition.ParentID)) {
            throw new Error('Invalid ParentID: ParentID must be a valid number');
        }
        try {
            const response = await axios.post(`${API_URL}/petitions`, petition, getAuthHeader());
            return response.data;
        } catch (error) {
            if (error.response?.status === 401) {
                await authService.refreshToken();
                const response = await axios.post(`${API_URL}/petitions`, petition, getAuthHeader());
                return response.data;
            }
            throw error.response?.data?.detail || error.message || 'Failed to create petition';
        }
    },

    /**
     * Cập nhật trạng thái đơn thỉnh cầu
     * @param {number} petitionId - ID của đơn thỉnh cầu
     * @param {Object} update - Thông tin cập nhật
     * @param {string} update.Status - Trạng thái mới
     * @param {string} update.Notes - Ghi chú/lý do
     * @returns {Promise<Object>} - Đơn thỉnh cầu đã được cập nhật
     */
    updatePetitionStatus: async (petitionId, update) => {
        if (!petitionId || isNaN(petitionId)) {
            throw new Error('Invalid petitionId: petitionId must be a valid number');
        }
        try {
            const response = await axios.put(`${API_URL}/petitions/${petitionId}`, update, getAuthHeader());
            return response.data;
        } catch (error) {
            if (error.response?.status === 401) {
                await authService.refreshToken();
                const response = await axios.put(`${API_URL}/petitions/${petitionId}`, update, getAuthHeader());
                return response.data;
            }
            throw error.response?.data?.detail || error.message || 'Failed to update petition status';
        }
    },

    /**
     * Lấy thống kê đơn thỉnh cầu
     * @param {Object} filters - Các bộ lọc
     * @param {string} filters.startDate - Ngày bắt đầu
     * @param {string} filters.endDate - Ngày kết thúc
     * @returns {Promise<Object>} - Thống kê số lượng đơn theo trạng thái
     */
    getPetitionStatistics: async (filters = {}) => {
        try {
            const response = await axios.get(`${API_URL}/petitions/statistics/`, {
                ...getAuthHeader(),
                params: filters
            });
            return response.data;
        } catch (error) {
            if (error.response?.status === 401) {
                await authService.refreshToken();
                const response = await axios.get(`${API_URL}/petitions/statistics/`, {
                    ...getAuthHeader(),
                    params: filters
                });
                return response.data;
            }
            throw error.response?.data?.detail || error.message || 'Failed to fetch petition statistics';
        }
    }
};

export default petitionService;

File: backend/app/services/petition_service.py
from sqlalchemy.orm import Session, joinedload
from ..models import Petition, User, Parent
from ..models.petition import PetitionStatus
from ..schemas.petition import PetitionCreate, PetitionUpdate
from typing import List, Optional
from datetime import datetime
from sqlalchemy import func
import logging

# Thiết lập logging để debug
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

class PetitionService:
    @staticmethod
    def get_petitions_by_parent(
        db: Session,
        parent_id: int,
        page: int = 1,
        size: int = 10
    ) -> tuple[List[Petition], int]:
        """
        Lấy danh sách đơn thỉnh cầu của một phụ huynh cụ thể
        
        Args:
            db (Session): Database session
            parent_id (int): ID của phụ huynh
            page (int): Số trang (mặc định là 1)
            size (int): Số lượng đơn trên mỗi trang (mặc định là 10)
            
        Returns:
            tuple[List[Petition], int]: Danh sách đơn và tổng số đơn
        """
        # Kiểm tra sự tồn tại của parent
        parent = db.query(Parent).filter(Parent.ParentID == parent_id).first()
        if not parent:
            raise ValueError(f"Parent with ID {parent_id} not found")

        # Eager load relationship parent và parent.user
        query = db.query(Petition).options(
            joinedload(Petition.parent).joinedload(Parent.user)
        ).filter(Petition.ParentID == parent_id)
        total = query.count()
        petitions = query.offset((page - 1) * size).limit(size).all()
        return petitions, total

    @staticmethod
    def get_all_petitions(
        db: Session,
        status: Optional[PetitionStatus] = None,
        parent_id: Optional[int] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        page: int = 1,
        size: int = 10
    ) -> tuple[List[Petition], int]:
        """
        Lấy tất cả đơn thỉnh cầu với các bộ lọc tùy chọn
        
        Args:
            db (Session): Database session
            status (Optional[PetitionStatus]): Lọc theo trạng thái đơn
            parent_id (Optional[int]): Lọc theo ID phụ huynh
            start_date (Optional[datetime]): Lọc theo ngày bắt đầu
            end_date (Optional[datetime]): Lọc theo ngày kết thúc
            page (int): Số trang (mặc định là 1)
            size (int): Số lượng đơn trên mỗi trang (mặc định là 10)
            
        Returns:
            tuple[List[Petition], int]: Danh sách đơn và tổng số đơn
        """
        query = db.query(Petition).options(
            joinedload(Petition.parent).joinedload(Parent.user)
        )
        
        if status:
            query = query.filter(Petition.Status == status)
        if parent_id:
            query = query.filter(Petition.ParentID == parent_id)
        if start_date:
            query = query.filter(Petition.SubmittedAt >= start_date)
        if end_date:
            query = query.filter(Petition.SubmittedAt <= end_date)
            
        total = query.count()
        petitions = query.offset((page - 1) * size).limit(size).all()
        return petitions, total

    @staticmethod
    def get_petition_by_id(db: Session, petition_id: int) -> Optional[Petition]:
        """
        Lấy thông tin chi tiết của một đơn thỉnh cầu theo ID
        
        Args:
            db (Session): Database session
            petition_id (int): ID của đơn thỉnh cầu
            
        Returns:
            Optional[Petition]: Đơn thỉnh cầu nếu tìm thấy, None nếu không tìm thấy
        """
        return db.query(Petition).options(
            joinedload(Petition.parent).joinedload(Parent.user)
        ).filter(Petition.PetitionID == petition_id).first()

    @staticmethod
    def create_petition(
        db: Session,
        petition: PetitionCreate,
        parent_id: int
    ) -> Petition:
        """
        Tạo một đơn thỉnh cầu mới
        """
        try:
            # Kiểm tra parent có tồn tại không
            parent = db.query(Parent).filter(Parent.ParentID == parent_id).first()
            if not parent:
                raise ValueError(f"Parent with ID {parent_id} not found")

            db_petition = Petition(
                ParentID=parent_id,
                Title=petition.Title,
                Content=petition.Content,
                Status=PetitionStatus.PENDING,
                SubmittedAt=datetime.utcnow()
            )
            db.add(db_petition)
            db.commit()
            db.refresh(db_petition)
            return db_petition
        except Exception as e:
            db.rollback()
            raise e

    @staticmethod
    def update_petition_status(
        db: Session,
        petition_id: int,
        update: PetitionUpdate,
        admin_id: int
    ) -> Optional[Petition]:
        """
        Cập nhật trạng thái và thông tin của một đơn thỉnh cầu
        
        Args:
            db (Session): Database session
            petition_id (int): ID của đơn thỉnh cầu cần cập nhật
            update (PetitionUpdate): Thông tin cập nhật
            admin_id (int): ID của người quản lý
            
        Returns:
            Optional[Petition]: Đơn thỉnh cầu đã được cập nhật nếu tìm thấy, None nếu không tìm thấy
        """
        petition = db.query(Petition).options(
            joinedload(Petition.parent).joinedload(Parent.user)
        ).filter(Petition.PetitionID == petition_id).first()
        
        if not petition:
            return None
            
        petition.Status = update.Status
        petition.AdminID = admin_id
        petition.Notes = update.Notes
        db.commit()
        db.refresh(petition)
        return petition

    @staticmethod
    def get_petition_statistics(
        db: Session,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None
    ) -> dict:
        """
        Retrieve statistics of petitions by status.

        Args:
            db (Session): Database session
            start_date (Optional[datetime]): Start date for filtering
            end_date (Optional[datetime]): End date for filtering

        Returns:
            dict: Dictionary with count of petitions by status (PENDING, APPROVED, REJECTED)
        """
        logger.debug(f"Querying petition statistics with start_date={start_date}, end_date={end_date}")

        query = db.query(
            Petition.Status,
            func.count(Petition.PetitionID).label('count')
        ).group_by(Petition.Status)

        if start_date:
            logger.debug(f"Filtering SubmittedAt >= {start_date}")
            query = query.filter(Petition.SubmittedAt >= start_date)
        if end_date:
            logger.debug(f"Filtering SubmittedAt <= {end_date}")
            query = query.filter(Petition.SubmittedAt <= end_date)

        results = query.all()
        logger.debug(f"Query results: {results}")

        # Initialize counts for all statuses using string keys
        stats = {
            "PENDING": 0,
            "APPROVED": 0,
            "REJECTED": 0
        }

        # Update counts from query results
        for status, count in results:
            status_str = status.value  # Lấy giá trị chuỗi của enum
            if status_str in stats:
                stats[status_str] = count
            else:
                logger.warning(f"Unexpected status: {status_str}")

        logger.debug(f"Final stats: {stats}")
        return stats